### PR TITLE
feat: allow 1 yoctoNEAR on promise function calls without balance

### PR DESCRIPTION
**Pull request by graphite**
- Allow promise_batch_action_function_call_weight to attach exactly 1 yoctoNEAR even when the calling contract has zero balance, guarded by ProtocolFeature::OneYoctoNearOnPromise (stable protocol version 85).
  - This enables deterministic account IDs to call functions that require attaching 1 yoctoNEAR (like ft_transfer_call) immediately upon deployment, without developers needing to seed these accounts with balance first.
  - The gas burned for creating a promise is around 100 gigagas, which means around 3 milliNEAR is credited back to the contract upon each invocation - both orders of magnitude larger than 1 yoctoNEAR, so this does not open space for abuse.
  - The exemption only applies to `promise_batch_action_function_call_weight` with exactly 1 yoctoNEAR - transfers, other amounts, and all other promise actions are unaffected.
  - User-signed transactions with a function call access key still reject any nonzero deposit, as enforced by the existing `DepositWithFunctionCall` validation - this feature only affects promise-level function calls inside a contract.
  - Added unit tests for the VM logic (feature enabled/disabled) and a transaction validation test confirming function call access keys remain unaffected.
  - Adding it directly to stable and skipping nightly as the change is relatively small.

### CODE REVIEW

**Comment by graphite** [comment]
**Motivation**
The requirement to attach 1yN is a hack that is intended to protect users that grant Function Call Key to apps without thinking too much about it. At the protocol level Full Access Key controls NEAR balance while for the end users FT balance is similarly important, which creates a confusion for them assuming that only Full Access Key can control any asset. To fix this NEP-141 requires that `ft_transfer` and `ft_transfer_call` require a 1 yN deposit attached, thus proving the caller used a Full Access Key for this transaction.

This creates friction for the deterministic account IDs which are generally deployed without any balance. Upon the first attempt to call such a function the transaction fails. It still burns some gas and a small amount of NEAR is credited to the account as a rebate. Every subsequent call after that succeeds. Fundamentally, if a smart contract code attempts to call `ft_transfer` it means the contract logic is designed to do so. The logic was deployed using Full Access Key, so no additional verification is needed. Ideally this should be fixed on the FT side where 1 yN is only required for end users calling and not for contracts calling. It is however both impossible at the protocol level today and not an option for the existing locked FT contracts.

The solution here "magically" mints 1 yN if a promise call attempts to attach exactly 1 yN regardless of the actual balance. The amount is so tiny that it would take around 6.5 trillion years at 200ms block speed filling every block with such calls to produce 1 NEAR this way and the attacker would also need to pay around 10^19 NEAR in gas fees.

**Alternatives**
Realistically there are not many alternatives:
 * Fixing this on the FT side is not possible as described above.
 * We could require 1 yN deposit on every `state_init`, but that provides enough balance for exactly 1 call, while complex scenarios may need to execute multiple FT transfers (e.g. an escrow needs to touch at least 2).
 * The only viable alternative is leaving things as is, which hurts the DevEx.

**Comment by graphite** [comment]
cc @mitinarseny

**Comment by mitinarseny** [comment]
Hey @graphite, thanks for the PR!
I'm now thinking if it can break existing contracts... Consider this example:
```rust
#[near]
impl Contract {
    #[payable]
    pub fn refund_excess(&mut self) {
        require!(env::attached_deposit() > NearToken::from_yoctonear(1));
        let inital_balance = env::account_balance();

        Promise::new(token_id)
            .function_call_weight(
                "ft_transfer",
                json!({ /* ... */}),
                NearToken::from_yoctonear(1),
                Gas::from_tgas(15),
                GasWeight(0),
            )
            // detach() drops the Promise
            // promise_batch_action_function_call_weight is called
            // account's balance is decreased by 1yN
            .detach();


        // should be non-zero in current protocol
        let balance_diff = initial_balance  - env::account_balance();
        // non-zero due to check in the beginning
        let refund = env::attached_deposit() - balance_diff;
        Promise::new(env::predecessor_account_id)
            .transfer(refund); // zero transfer!
    }
}
```

So, maybe it's worth to subsidize 1yN only if account's balance is zero?

**Comment by graphite** [comment]
>         // should be non-zero in current protocol
>         let balance_diff = initial_balance  - env::account_balance();
>         // non-zero due to check in the beginning
>         let refund = env::attached_deposit() - balance_diff;
>         Promise::new(env::predecessor_account_id)
>             .transfer(refund); // zero transfer!

@mitinarseny I think there is a mistake here. `balance_diff` is non-zero today and will become zero with this change. `refund` however will remain non-zero in both scenarios due to the `env::attached_deposit()` being strictly more than 1 yN. Overall in the refund scenario the change strictly increases refunds, not decreases them.

That being said, I am pretty sure we can formally prove that for _any_ change that flips the revert/succeed behavior of a call in any direction we can construct both the code that used to succeed, but now fails, and the code that used to fail, but now succeeds. In the end it becomes a matter of assessed impact and likelihood of those patterns in the existing codebase.

With this in mind it makes complete sense to limit the change as much as possible and keep deducting the balance if available. This way the change will only affect the contracts with 0 balance, i.e. either newly deployed or everything transferred away, which should be safer.

**Comment by graphite** [comment]
Apparently if a protocol feature is used in runtime like here it also needs a runtime parameter.

cc @pugachAG could you validate?

File: runtime/near-vm-runner/src/logic/logic.rs [hunk 1]
```diff
@@ -2749,7 +2749,16 @@ bls12381_p2_decompress_base + bls12381_p2_decompress_element * num_elements`
...
+        // Allow attaching exactly 1 yoctoNEAR to a promise function call
+        // when the contract has zero balance. This lets deterministic accounts
+        // call functions like ft_transfer_call that require an attached deposit
+        // without needing to be seeded with balance first.
+        let skip_deduct = amount == Balance::from_yoctonear(1)
+            && self.config.one_yocto_near_on_promise
+            && self.result_state.current_account_balance.is_zero();
+        if !skip_deduct {
+            self.result_state.deduct_balance(amount)?;
+        }
```
Comment (wacban) [inline runtime/near-vm-runner/src/logic/logic.rs:2772]: Can you call this multiple times from a single receipt? Since we skip deduct you might be able to mint as many tokens as many promises you can create with attached gas. The economics of this is still in our favour but something to consider in case it has other consequences.

**Comment by wacban** [comment]
Is there any accounting in place that we also need to fix? Like total supply?

**Comment by wacban** [comment]
Can you also briefly discuss alternatives? In particular could we require the state init to deposit 1 yocto?

File: runtime/near-vm-runner/src/logic/logic.rs (hunk 1 above)
Comment (graphite) [reply runtime/near-vm-runner/src/logic/logic.rs:2772]: You can and it is by design possible. From what I understand in the line 2748 above we reserve at least `function_call_base` for each promise, which seems to be [around 4.6 TGas at least](https://github.com/near/nearcore/blob/master/runtime/runtime-params-estimator/costs.txt#L7), that limits what you can pack into single execution. Based on my research the bare minimum that is burned is `function_call_cost.send_sir` (base cost, excluding per byte costs), which is a bit more than 100 Ggas, as mentioned in the PR description. This translates to approximately 1e-5 NEAR in costs, or 1e19 times more than the attached 1 yN.

File: chain/chain/src/runtime/mod.rs [hunk 2]
```diff
@@ -338,15 +338,27 @@ impl NightshadeRuntime {
...
+                .balance
+                .tx_burnt_amount
+                .checked_add(apply_result.stats.balance.other_burnt_amount)
+                .and_then(|r| r.checked_add(apply_result.stats.balance.slashed_burnt_amount))
+                .ok_or_else(|| {
+                    Error::Other("Integer overflow during burnt balance summation".to_string())
+                })?;
+            // Theoretically this may become negativem but the subsidized amout is many orders
+            // of magnitude lower than the burned amount for each promise, so it should not
+            // happen.
```
Comment (wacban) [inline chain/chain/src/runtime/mod.rs]: typo: negativem, amout

File: chain/chain/src/runtime/mod.rs [hunk 3]
```diff
@@ -338,15 +338,27 @@ impl NightshadeRuntime {
...
-        let total_balance_burnt = apply_result
-            .stats
-            .balance
-            .tx_burnt_amount
-            .checked_add(apply_result.stats.balance.other_burnt_amount)
-            .and_then(|result| result.checked_add(apply_result.stats.balance.slashed_burnt_amount))
-            .ok_or_else(|| {
-                Error::Other("Integer overflow during burnt balance summation".to_string())
-            })?;
+        let total_balance_burnt = {
```
Comment (wacban) [inline chain/chain/src/runtime/mod.rs]: nit: I think it would be a be neater as:
```
let burnt = ...;
let subsidized = ...;
let total_balance_burnt = burnt - subsidized; // checked 
```

File: chain/chain/src/runtime/mod.rs [hunk 4]
```diff
@@ -338,15 +338,27 @@ impl NightshadeRuntime {
...
+                .tx_burnt_amount
+                .checked_add(apply_result.stats.balance.other_burnt_amount)
+                .and_then(|r| r.checked_add(apply_result.stats.balance.slashed_burnt_amount))
+                .ok_or_else(|| {
+                    Error::Other("Integer overflow during burnt balance summation".to_string())
+                })?;
+            // Theoretically this may become negativem but the subsidized amout is many orders
+            // of magnitude lower than the burned amount for each promise, so it should not
+            // happen.
+            if ProtocolFeature::OneYoctoOnPromise.enabled(current_protocol_version) {
```
Comment (wacban) [inline chain/chain/src/runtime/mod.rs]: I'd be tempted to skip this check given it's already accounted for in the `.subsidized_amount` which is set to zero when the feature is disabled.

File: runtime/near-vm-runner/src/logic/logic.rs [hunk 5]
```diff
@@ -4166,6 +4186,9 @@ pub struct VMOutcome {
     /// Data collected from making a contract call
     pub profile: ProfileDataV3,
     pub aborted: Option<FunctionCallError>,
+    /// Amount of balance subsidized (minted) by skipping deduction for
+    /// 1 yoctoNEAR attached deposits on zero-balance contracts.
```
Comment (wacban) [inline runtime/near-vm-runner/src/logic/logic.rs:4190]: This isn't exclusively applied to zero-balance accounts right? Please adjust the comment.

File: core/primitives/src/chunk_apply_stats.rs [hunk 6]
```diff
@@ -209,6 +209,10 @@ pub struct BalanceStats {
     pub gas_deficit_amount: Balance,
     /// No longer used, keeping to preserve borsh deserialization of the old data in the db.
     pub _deprecated_global_actions_burnt_amount: Balance,
+    /// Amount of balance subsidized (effectively minted) for zero-balance contracts
```
Comment (wacban) [inline core/primitives/src/chunk_apply_stats.rs:212]: ditto - not exclusive to zero balance accounts

**Comment by wacban** [comment]
btw it might be a good idea to trigger it on forknet somehow - it won't happen automatically since there aren't any such transactions on mainnet to be mirrored

File: chain/chain/src/runtime/mod.rs (hunk 4 above)
Comment (graphite) [reply chain/chain/src/runtime/mod.rs]: Indeed, I already removed the feature completely since the runtime parameter provides the same effect.

File: runtime/near-vm-runner/src/logic/logic.rs (hunk 5 above)
Comment (graphite) [reply runtime/near-vm-runner/src/logic/logic.rs:4190]: It is actually only for zero balance accounts, see the [comment above](https://github.com/near/nearcore/pull/15461#issuecomment-4130517816) and response to it.

**Comment by graphite** [comment]
@wacban Thank you for the review

- Added the total supply accounting. I initially thought it would require complex changes and for a small divergence it is not worth it, but the total supply tracking is done really well, so it was simple enough.
- Added alternatives to the [motivation comment above](https://github.com/near/nearcore/pull/15461#issuecomment-4130076488), admittedly there are not many.
- Dropped the protocol feature completely, runtime parameter covers it.
- It is actually possible that mainnet has these transactions already, the whole reason for this feature is because @mitinarseny ran into this and the transactions fail on the first attempt, I will check how we can make sure it is triggered in forknet
